microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/internal/openapi/java/MicroProfileGenerateOpenAPIOperation.java

```diff
@@ -0,0 +1,69 @@
+/*******************************************************************************
+* Copyright (c) 2020 Red Hat Inc. and others.
+* All rights reserved. This program and the accompanying materials
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v20.html
+*
+* SPDX-License-Identifier: EPL-2.0
+*
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package com.redhat.microprofile.jdt.internal.openapi.java;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.dom.CompilationUnit;
+import org.eclipse.jdt.core.dom.TypeDeclaration;
+import org.eclipse.lsp4j.CodeAction;
+import org.eclipse.lsp4j.Diagnostic;
+
+import com.redhat.microprofile.jdt.core.java.codeaction.IJavaCodeActionParticipant;
+import com.redhat.microprofile.jdt.core.java.codeaction.JavaCodeActionContext;
+import com.redhat.microprofile.jdt.core.java.corrections.proposal.ChangeCorrectionProposal;
+import com.redhat.microprofile.jdt.core.utils.JDTTypeUtils;
+import com.redhat.microprofile.jdt.internal.openapi.MicroProfileOpenAPIConstants;
+
+/**
+ * Generate OpenAPI annotations by the "Source" kind code action.
+ * 
+ * @author Benson Ning
+ *
+ */
+public class MicroProfileGenerateOpenAPIOperation implements IJavaCodeActionParticipant {
+	
+	@Override
+	public boolean isAdaptedForCodeAction(JavaCodeActionContext context, IProgressMonitor monitor)
+			throws CoreException {
+		IJavaProject javaProject = context.getJavaProject();
+		return JDTTypeUtils.findType(javaProject, MicroProfileOpenAPIConstants.OPERATION_ANNOTATION) != null;
+	}
+
+	@Override
+	public List<? extends CodeAction> getCodeActions(JavaCodeActionContext context, Diagnostic diagnostic,
+			IProgressMonitor monitor) throws CoreException {
+		List<CodeAction> codeActions = new ArrayList<>();
+		CompilationUnit cu = context.getASTRoot();
+		List<?> types = cu.types();
+		for (Object type : types){
+			if (type instanceof TypeDeclaration) {
+				ChangeCorrectionProposal proposal = new OpenAPIAnnotationProposal(
+						"Generate OpenAPI Annotations", context.getCompilationUnit(), context.getASTRoot(),
+						(TypeDeclaration) type, MicroProfileOpenAPIConstants.OPERATION_ANNOTATION, 0);
+				// Convert the proposal to LSP4J CodeAction
+				CodeAction codeAction = context.convertToCodeAction(proposal);
+				if (codeAction == null) {
```

Comment:
Code Action should be added only if there are some changes, no? At first I tried to execute this CodeAction into the ExampleResource:

```
@Path("/hello")
public class ExampleResource {

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    public String hello() {
        return "hello";
    }
}
```

and it does nothing. After debugging the code action, the problem occurs because context.convertToCodeAction(proposal); return null. I think code action should be added only if code action does something, no?

A second question, is it normal that Operation cannot be generated from a simple hello service which returns a String (instead of Response)?

Reply:
1. Talked to @ericglau about the behaviour before, he prefers to show the code action even there is nothing to generate to the Java file
2. @ericglau ^^

Reply:
M'y first feeling has been that thé code action didnt work because it generates something. But if you prefer having this behavior please add a comment to explain why you add an empty code action. @fbricon what do you think about this behavior? Do you provide thé code action in jdt les if it dors nothing?

Reply:
Yeah I strongly disagree with showing a useless code action. It should only show when it's useful